pairing/bls12381/point.go

```diff
@@ -0,0 +1,403 @@
+package bls12381
+
+import (
+	"crypto/cipher"
+	"crypto/sha256"
+	"encoding/hex"
+	"io"
+
+	bls12381 "github.com/kilic/bls12-381"
+	"go.dedis.ch/kyber/v3"
+	"go.dedis.ch/kyber/v3/group/mod"
+)
+
+// pointG1 is a kyber.Point holding a G1 point on BLS12-381 curve
+type pointG1 struct {
+	p *bls12381.PointG1
+}
+
+func nullpointG1() *pointG1 {
+	var p bls12381.PointG1
+	return newPointG1(&p)
+}
+
+func newPointG1(p *bls12381.PointG1) *pointG1 {
+	return &pointG1{p: p}
+}
+
+func (k *pointG1) Equal(k2 kyber.Point) bool {
+	return bls12381.NewG1().Equal(k.p, k2.(*pointG1).p)
+}
+
+func (k *pointG1) Null() kyber.Point {
+	return newPointG1(bls12381.NewG1().Zero())
+}
+
+func (k *pointG1) Base() kyber.Point {
+	return newPointG1(bls12381.NewG1().One())
+}
+
+func (k *pointG1) Pick(rand cipher.Stream) kyber.Point {
+	var dst, src [32]byte
+	rand.XORKeyStream(dst[:], src[:])
+	return k.Hash(dst[:])
+}
+
+func (k *pointG1) Set(q kyber.Point) kyber.Point {
+	k.p.Set(q.(*pointG1).p)
+	return k
+}
+
+func (k *pointG1) Clone() kyber.Point {
+	var p bls12381.PointG1
+	p.Set(k.p)
+	return newPointG1(&p)
+}
+
+func (k *pointG1) EmbedLen() int {
+	panic("bls12-381: unsupported operation")
+}
+
+func (k *pointG1) Embed(data []byte, rand cipher.Stream) kyber.Point {
+	panic("bls12-381: unsupported operation")
+}
+
+func (k *pointG1) Data() ([]byte, error) {
+	panic("bls12-381: unsupported operation")
+}
+
+func (k *pointG1) Add(a, b kyber.Point) kyber.Point {
+	aa := a.(*pointG1)
+	bb := b.(*pointG1)
+	bls12381.NewG1().Add(k.p, aa.p, bb.p)
+	return k
+}
+
+func (k *pointG1) Sub(a, b kyber.Point) kyber.Point {
+	aa := a.(*pointG1)
+	bb := b.(*pointG1)
+	bls12381.NewG1().Sub(k.p, aa.p, bb.p)
+	return k
+}
+
+func (k *pointG1) Neg(a kyber.Point) kyber.Point {
+	aa := a.(*pointG1)
+	bls12381.NewG1().Neg(k.p, aa.p)
+	return k
+}
+
+func (k *pointG1) Mul(s kyber.Scalar, q kyber.Point) kyber.Point {
+	if q == nil {
+		q = nullpointG1().Base()
+	}
+	bls12381.NewG1().MulScalarBig(
+		k.p,
+		q.(*pointG1).p,
+		&s.(*mod.Int).V,
+	)
+	return k
+}
+
+func (k *pointG1) MarshalBinary() ([]byte, error) {
+	return bls12381.NewG1().ToCompressed(k.p), nil
+}
+
+func (k *pointG1) UnmarshalBinary(buff []byte) error {
+	var err error
+	k.p, err = bls12381.NewG1().FromCompressed(buff)
+	return err
+}
+
+func (k *pointG1) MarshalTo(w io.Writer) (int, error) {
+	buf, err := k.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	return w.Write(buf)
+}
+
+func (k *pointG1) UnmarshalFrom(r io.Reader) (int, error) {
+	buf := make([]byte, k.MarshalSize())
+	n, err := io.ReadFull(r, buf)
+	if err != nil {
+		return n, err
+	}
+	return n, k.UnmarshalBinary(buf)
+}
+
+func (k *pointG1) MarshalSize() int {
+	return 48
+}
+
+func (k *pointG1) String() string {
+	b, _ := k.MarshalBinary()
+	return "bls12-381.G1: " + hex.EncodeToString(b)
+}
+
+func (k *pointG1) Hash(m []byte) kyber.Point {
+	p, _ := bls12381.NewG1().HashToCurve(m, Domain)
```

Comment:
@ineiti should we `panic()` if there is an error ?

Reply:
Hmm - good question. I don't see the `Hash` defined in `/group.go`, so I don't know where the function signature comes from. 

But if it's not possible to return an error, and the point cannot be created, then I'd call `panic()` here, yes. Better than continuing with an invalid Point!

Reply:
The same function signature is used for `bn256`, so no error is returned. I would `panic` in case of error, as it is done in `bn256` to avoid carrying around an invalid point.